storage: reduce allocator thrashing

cuongdo opened the pull request:
**storage: reduce allocator thrashing**
Because the replica allocator is asynchronous and relies on gossipped
store descriptors that are up to 60 seconds old, thrashing has been
observed in production and the 3->5 allocator test. This fixes that by
only considering rebalancing if

   stddev(replica counts) >=
     max(0.5 \* mean(replica counts), 1.5)

The 1.5 was derived through experiments and found to work well with very
small data (15 replicas total) and moderate-sized data (thousands of
replicas).

Fixes #9031 and #8761

**terrafarm: always quote Terraform values** (for testing other commit)
When setting Terraform variables, Farmer was only quoting the values of
two variables.  Now, we quote all values. This resolves the issue where
passed in TESTFLAGS (e.g. `-tf.cockroach-binary`) were being silently
ignored when passed into Terraform because of some unexpected behavior
when using implicitly typed Terraform values with Terraform functions
such as `coalesce`.

For example, when specifying `-tf.cockroach-binary`, Farmer passes
`-var=cockroach_binary=foo` to Terraform. `main.tf` does this to figure
out which `cockroach` binary to copy to the new nodes:

```
coalesce(cockroach_binary, ...)
```

`coalesce` is supposed to return the first non-empty string passed to
it. When `cockroach_binary` is an unquoted string, `coalesce` is
silently ignores the non-empty value for `cockroach_binary`.

cc @vivekmenezes (since you were asking about this)

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9216)

<!-- Reviewable:end -->

Commit:
terrafarm: always quote Terraform values

When setting Terraform variables, Farmer was only quoting the values of
two variables.  Now, we quote all values. This resolves the issue where
passed in TESTFLAGS (e.g. `-tf.cockroach-binary`) were being silently
ignored when passed into Terraform because of some unexpected behavior
when using implicitly typed Terraform values with Terraform functions
such as `coalesce`.

For example, when specifying `-tf.cockroach-binary`, Farmer passes
`-var=cockroach_binary=foo` to Terraform. `main.tf` does this to figure
out which `cockroach` binary to copy to the new nodes:

```
coalesce(cockroach_binary, ...)
```

`coalesce` is supposed to return the first non-empty string passed to
it. When `cockroach_binary` is an unquoted string, `coalesce` is
silently ignores the non-empty value for `cockroach_binary`.

Commit:
storage: reduce allocator thrashing

Because the replica allocator is asynchronous and relies on gossipped
store descriptors that are up to 60 seconds old, thrashing has been
observed in production and the 3->5 allocator test. This fixes that by
only considering rebalancing if

   stddev(replica counts) >=
     max(0.5 * mean(replica counts), 1.5)

The 1.5 was derived through experiments and found to work well with very
small data (15 replicas total) and moderate-sized data (thousands of
replicas).

Fixes #9031
Fixes #8761

tamird commented:
FYI "Fixes #9031 and #8761" won't do what you want. You'll want to write it as:

Fixes #9031.
Fixes #8761.

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>, just some style commentary.

---

Reviewed 2 of 2 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[acceptance/util_test.go, line 110 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRAASEUgwNb2EO84X1M:-KRAASEUgwNb2EO84X1N:b-khpbhs) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/acceptance/util_test.go#L110)):*
> ```Go
> // Allocator test flags.
> var flagATMaxStdDev = flag.Float64("at.std-dev", 0,
> 	"maximum standard deviation of replica counts (default: 5% of mean replica count)")
> ```

do you need to explain the default in the usage string? since the default value is passed to `flag.Float64`, I'd expect that it (the default) would already be included in the string representation of the flag. Is that not the case?

---

*[storage/allocator.go, line 42 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRABffFu228hOqdrqIc:-KRABffFu228hOqdrqId:b-dl7wjf) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator.go#L42)):*
> ```Go
> 	maxFractionUsedThreshold = 0.95
> 
> 	// MinStdDevPercentForRebalance is the minimum percentage of the mean replica
> ```

"percentage" isn't right - this value is a multiple. might want to update the name as well.

this is also quite a hard sentence to grok. Perhaps "minimum ratio of the standard deviation of per-store replica counts to the mean of per-store replica counts"

---

*[storage/allocator.go, line 46 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRACWn4HnNENInxDBCv:-KRACWn4HnNENInxDBCw:bdjr36f) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator.go#L46)):*
> ```Go
> 	// is possible when:
> 	//
> 	// stddev(rangeCounts) >= max(minStdDevPercentForRebalance * meanRangeCount,
> ```

mean(rangeCounts) instead of meanRangeCount?

---

*[storage/allocator_test.go, line 469 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRAD3BEe1F41fZ39kzX:-KRAD3BEe1F41fZ39kzY:b312n4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator_test.go#L469)):*
> ```Go
> 	defer stopper.Stop()
> 
> 	// Stores are almost balanced => should not rebalance
> ```

What is this comment about? There are no range counts on these stores yet.

---

*[storage/allocator_test.go, line 470 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRADbc4eBeXuseSU8_5:-KRADbc4eBeXuseSU8_6:b-c4tiwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator_test.go#L470)):*
> ```Go
> 
> 	// Stores are almost balanced => should not rebalance
> 	stores := []*roachpb.StoreDescriptor{
> ```

nit: all the slices in this test might want to be fixed-size arrays of 3 elements - otherwise it is worryingly easy to muck things up here if we're not careful in the future.

---

*[storage/allocator_test.go, line 488 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRADDfsn73vxB2-mnBv:-KRADDfsn73vxB2-mnBw:bmn2h2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator_test.go#L488)):*
> ```Go
> 	}
> 
> 	belowRebal := func(n int32) int32 {
> ```

what is this function doing? where are the values it is called with coming from?

---

*[storage/allocator_test.go, line 508 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRADn7ZFwqhYAzQ7aS7:-KRADn7ZFwqhYAzQ7aS8:b8unmcx) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator_test.go#L508)):*
> ```Go
> 		{[]int32{10, 8, 9}, []bool{false, false, false}},
> 
> 		{[]int32{belowRebal(100) + 1, 100, aboveRebal(100) - 1}, []bool{false, false, false}},
> ```

see comment above about magic numbers. a sprinkling of commentary (or usefully named constants) would be a great helper here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-:-KRABHgqzoYmm3wRpfLz:buytgf1)*
<!-- Sent from Reviewable.io -->

petermattis commented:
The range counts are now returned via reservation calls, so the allocator information should be much more up to date than 60 seconds. One asymmetry here is that a node which receives a range now has older information than the one which sent the reservation. Perhaps we should be exchanging the range counts both directions when we performing rebalancing.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/allocator.go, line 387 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRAC1Wfcz86C7fxe3F6:-KRAC1Wfcz86C7fxe3F7:b-6x9enx) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator.go#L387)):*
> ```Go
> 	}
> 	stdDev := sl.count.stddev()
> 	if stdDev < minStdDev {
> ```

I don't have an intuition for when this stops rebalancing. I think it is worth some examples here showing what the stddev is and what action the allocator will take. For example, if there are 3 nodes containing 7 ranges and a new empty node is added we'll get:

```
  7,7,7,0 stddev=3.5
  7,7,6,1 stddev=2.8
  7,6,6,2 stddev=2.2
  6,6,6,3 stddev=1.5
  6,6,5,4 stddev=0.9
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-:-KRADuJzcXiWuIt2gMaE:b-cr3d70)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
WIP to address some of Tamir's comments and Peter's comment. I think I'll need to rework the new test to make it more obvious and readable.

More changes to come.

---

Review status: 4 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

*[acceptance/util_test.go, line 110 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRAASEUgwNb2EO84X1M:-KRAFnM9wvId5MDuwSDM:bxn51nm) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/acceptance/util_test.go#L110)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> do you need to explain the default in the usage string? since the default value is passed to `flag.Float64`, I'd expect that it (the default) would already be included in the string representation of the flag. Is that not the case?
</details>
The default value is 0, and it's not obvious what the semantics of a 0 value for this flag are.

---

*[storage/allocator.go, line 42 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRABffFu228hOqdrqIc:-KRAGEbGG55EvaSiBgcK:b-usyy17) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator.go#L42)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "percentage" isn't right - this value is a multiple. might want to update the name as well.
> 
> this is also quite a hard sentence to grok. Perhaps "minimum ratio of the standard deviation of per-store replica counts to the mean of per-store replica counts"
</details>
I like your use of "ratio" over "percentage", so I've changed it

---

*[storage/allocator.go, line 46 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRACWn4HnNENInxDBCv:-KRAI-9x0uCTBOCXL3un:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator.go#L46)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> mean(rangeCounts) instead of meanRangeCount?
</details>
Done.

---

*[storage/allocator.go, line 387 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRAC1Wfcz86C7fxe3F6:-KRAOFukIZwwrLZIHQuk:b-pprol9) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator.go#L387)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I don't have an intuition for when this stops rebalancing. I think it is worth some examples here showing what the stddev is and what action the allocator will take. For example, if there are 3 nodes containing 7 ranges and a new empty node is added we'll get:
> 
> ```
>   7,7,7,0 stddev=3.5
>   7,7,6,1 stddev=2.8
>   7,6,6,2 stddev=2.2
>   6,6,6,3 stddev=1.5
>   6,6,5,4 stddev=0.9
> ```
</details>
Added a large comment block

---

*[storage/allocator_test.go, line 469 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRAD3BEe1F41fZ39kzX:-KRAOLfOYJAjmnvO6VuV:bc7urf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator_test.go#L469)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> What is this comment about? There are no range counts on these stores yet.
</details>
from a previous rev; removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-:-KRAPwz7Ks1VlYwFg3Dz:bm0yxk9)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[acceptance/allocator_test.go, line 376 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRAShYPC_AQUm2I8-o3:-KRAShYPC_AQUm2I8-o4:bw9r9t1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c70e61a6712ae3e509ed4abd809ff0d6c6106576/acceptance/allocator_test.go#L376)):*
> ```Go
> 			if StableInterval <= stats.ElapsedSinceLastEvent {
> 				host := at.f.Nodes()[0]
> 				log.Infof(context.Background(), "replica count = %f, max = %f", stats.ReplicaCountStdDev, *flagATMaxStdDev)
> ```

s/*flagATMaxStdDev/maxStdDev/?

---

*[acceptance/util_test.go, line 110 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRAASEUgwNb2EO84X1M:-KRASxrPYE1qgjq9C3Uq:bemiy5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/acceptance/util_test.go#L110)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> The default value is 0, and it's not obvious what the semantics of a 0 value for this flag are.
</details>
Does this definition of the flag still make sense? ISTM that this flag should go away and the test should use the same criteria as the allocator.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Hi Cuong,

What feels strange to me is that the formula prevents rebalancing based on cluster-wide averages and variances, which can look ok in a big cluster even when there's only one or two outliers - and we definitely want to rebalance the outliers..

Some examples:
- 99 stores have 500 ranges, 1 store has 2000 ranges. mean=515, stddev=148, no go.
- 99 stores have 500 ranges, 1 store has nothing. mean=495, stddev=49, no go. This is realistic if we add or recover a node. This is a problem in most large clusters, even if they're not perfectly balanced - in a 100 node cluster, the new node will only decrease the mean by 1%, and will increase the standard deviation by at most `mean/10`. 

We can give more weight to outliers by using higher moments: `(sum(|x_i - mean|^k) / N)^(1/k)`. We are using  k=2. The higher `k` is, the more important outliers are; when `k -> infinity`, this is just the maximum difference `|x_i - mean|`. But `k` would be one more thing to tune..  Would just looking at the maximum difference (i.e. `k=inf`) work?

The 1.5 constant is also a bit dubious as it's not scale invariant (if I make all counts 10 times larger or smaller, should that change the decision whether to rebalance?)

If avoiding thrashing is the goal here, one possible mechanism is to have different thresholds for moving things TO and FROM a store (as part of rebalancing). E.g. we are only allowed to move ranges to stores that have at most `0.8*mean` ranges, and we are only allowed to move ranges from stores that have at least `1.2*mean` ranges. Of course, this won't help if the staleness of our counts is the real issue (but then nothing will work reliably).

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-:-KRAJqccKsbAZtX0GI2C:b-z8qukp)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Per @RaduBerinde: in a previous life the balancing heuristic we used was to rebalance from nodes that where fuller than 1.05x the mean and to nodes that were less than 0.95x the mean. 

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-:-KRAcJEwG49N8CqDxaxO:b-njz3f7)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
@RaduBerinde: the 1.5 is an absolute floor on the stddev needed to consider
rebalancing. It helps only for really small replica counts. As replica
counts get larger. The 0.5*mean calculation quickly renders the 1.5
irrelevant.
On Thu, Sep 8, 2016 at 5:10 PM Peter Mattis <notifications@github.com>
wrote:

> Per @RaduBerinde <https://github.com/RaduBerinde>: in a previous life the
> balancing heuristic we used was to rebalance from nodes that where fuller
> than 1.05x the mean and to nodes that were less than 0.95x the mean.
> ------------------------------
>
> Review status: all files reviewed at latest revision, 6 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-:-KRAcJEwG49N8CqDxaxO:b-njz3f7>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9216#issuecomment-245742232>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABffpmmGUInYmLD0ek3BuXgCzsdlzerLks5qoHnZgaJpZM4J4T4k>
> .
>

bdarnell commented:
When you say that 1.5 was "experimentally derived", please describe those experiments in sufficient detail that they can be reproduced and we can tell if it should be changed in the future. 

---

Reviewed 2 of 2 files at r1, 2 of 5 files at r2, 2 of 3 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/allocator.go, line 387 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-KRAC1Wfcz86C7fxe3F6:-KRCWMIQ1K8VzHRm4slx:b-is4jvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/974ebce583d4194cf61298b6f9c658130bd3d47d/storage/allocator.go#L387)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> Added a large comment block
</details>
The comment should also explain why the standard deviation is worth considering here. It seems to me that it's introducing a lot of unintuitive math for what should be fairly simple: imbalances should be larger than a single range before we take action. I was thinking that we'd solve this problem by setting [`replicaImbalanceTolerance`](https://github.com/cockroachdb/cockroach/blob/bad99359066aea31aef1ffce81fbdb6bd7a535b9/storage/allocator.go#L305) (in the `develop` version of this code, which has diverged a bit) to a value higher than 1, maybe `replicaImbalanceTolerance = max(meanRangeCount*0.05, 2)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9216#-:-KRCYb1g9m8cscXyE5Ee:brn0mgn)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
abandoning in favor of #9230

tamird commented on storage/allocator_test.go:
```diff
@@ -455,6 +457,102 @@ func TestAllocatorRebalance(t *testing.T) {
+// TestAllocatorRebalanceThrashing tests that the rebalancer does not thrash
+// when replica counts are balanced, within the appropriate thresholds, across
+// stores.
+func TestAllocatorRebalanceThrashing(t *testing.T) {
```
perhaps this test is worth bringing over?
